### ISSUE TITLE
Add new custom attributes to attributeGroup fails.

### ISSUE THREAD

**Issue by mattiaclementi** [magento/magento2#4461]
## Steps to reproduce
1. Install Magento from `2.0` branch.
2. Add a custom command to Magento's CLI called _example:module:create-feature_ or as you like;
3. The goal of this command is to create a series of custom **attributes**, connect this with a new custom **attributeGroup** and connect this attribute group with a new **attributeSet** that extend the native attribute set of Product entity;
   1. First create a attributeSet called "**Test**";
   2. Create a attributeGroup called _as you like_ and connect to the attributeSet called "Test";
   3. Create a set of attributes from an array from which to extract the names of these attributes, the array i set is like this: `protected $validColumnNames = ['CATEGORY_UID','BRAND_UID'];`
4. Start command from the CLI with this sintax _php bin/magento example:modules:create-feature_ or similiar.
## Expected result

After executing the command I expect to see in Magento admin interface:
1.  A new attributeSet called "**Test**"
2.  Some new custom attributes connect to attributeSet called "Test"
3. A new custom attributeGroup with the new custom attributes assign to itself.
## Actual result

After the executing of command Magento is modify but not correctly:
1. I see the new attirbuteSet called "Test".
2. I see the new attributeGroup created and assigned to "Test" attributeSet;
3. I see only the first new custom attribute assign to the attributeGroup in the new attributeSet, but the remaining attributes are created but not assigned to the attributeGroup;
4. The command ends but during the execution generates some exceptions while trying to assign attributes to attribute group;
## Actual Code

This is my actual command code:

```
<?php
namespace Qweb\ImportProducts\Console\Command;

use Symfony\Component\Config\Definition\Exception\Exception;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputArgument;
use Symfony\Component\Console\Input\InputOption;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Output\OutputInterface;

class CreateProductCommand extends Command
{
    /**
     * Object Manager
     *
     * @var \Magento\Framework\ObjectManager\ObjectManager
     */
    protected $objectManager;

    /**
     * Valid column names
     *
     * @array
     */
    protected $validColumnNames = ['CATEGORY_UID','BRAND_UID'];

    public function __construct(
        \Magento\Framework\ObjectManagerInterface $objectManager,
        \Magento\Catalog\Api\ProductRepositoryInterface $productRepositoryInterface,
        \Magento\Catalog\Api\Data\ProductInterface $productInterface,
        \Magento\Catalog\Api\CategoryManagementInterface $categoryManagementInterface,
        \Magento\Catalog\Api\CategoryLinkManagementInterface $categoryLinkManagementInterface,
        \Magento\Framework\App\State $state,
        \Magento\Eav\Api\Data\AttributeSetInterface $attributeSetInterface,
        \Magento\Eav\Api\AttributeSetRepositoryInterface $attributeSetRepositoryInterface,
        \Magento\Eav\Api\AttributeSetManagementInterface $attributeSetManagementInterface,
        \Magento\Eav\Api\Data\AttributeInterface $attributeInterface,
        \Magento\Eav\Api\Data\AttributeGroupInterface $attributeGroupInterface,
        \Magento\Eav\Api\AttributeGroupRepositoryInterface $attributeGroupRepositoryInterface,
        \Magento\Eav\Api\AttributeRepositoryInterface $attributeRepositoryInterface,
        \Magento\Eav\Api\AttributeManagementInterface $attributeManagementInterface)
    {
        parent::__construct();
        $this->objectManager = $objectManager;
        $this->productRepositoryInterface = $productRepositoryInterface;
        $this->productInterface = $productInterface;
        $this->categoryManagement = $categoryManagementInterface;
        $this->categoryLinkManagement = $categoryLinkManagementInterface;
        $this->attributeSetInterface = $attributeSetInterface;
        $this->attributeSetRepositoryInterface = $attributeSetRepositoryInterface;
        $this->attributeSetManagementInterface = $attributeSetManagementInterface;
        $this->attributeInterface = $attributeInterface;
        $this->attributeGroupInterface = $attributeGroupInterface;
        $this->attributeGroupRepositoryInterface = $attributeGroupRepositoryInterface;
        $this->attributeRepositoryInterface = $attributeRepositoryInterface;
        $this->attributeManagementInterface = $attributeManagementInterface;
        $state->setAreaCode('frontend');
    }

    /**
     * {@inheritdoc}
     */
    protected function configure()
    {
        $this->setName('example:modules:create-feature')
            ->setDescription('Create new custom attribute from array');

        parent::configure();
    }

    /**
     * {@inheritdoc}
     */
    protected function execute(InputInterface $input, OutputInterface $output)
    {
        $attributeSet = $this->attributeSetInterface;
        $attributeSet->setEntityTypeId(4);
        $attributeSet->setAttributeSetName("Test");

        try {
            $attributeSet = $this->attributeSetManagementInterface->create(4, $attributeSet, 4);
        } catch (\Exception $e) {
            var_dump($e->getFile());
            var_dump($e->getLine());
            var_dump($e->getTraceAsString());
            var_dump($e->getMessage());
        }

        $attributeSetRepository = $this->attributeSetRepositoryInterface;

        try {
            $attributeSetRepository->save($attributeSet);
        } catch (\Exception $e) {
            var_dump($e->getFile());
            var_dump($e->getLine());
            var_dump($e->getTraceAsString());
            var_dump($e->getMessage());
        }

        $attributeGroup = $this->attributeGroupInterface;
        $attributeGroup->setAttributeSetId($attributeSet->getAttributeSetId());
        $attributeGroup->setAttributeGroupId(null);
        $attributeGroup->setAttributeGroupName("Custom Attribute");
        $this->attributeGroupRepositoryInterface->save($this->attributeGroupInterface);


        foreach($this->validColumnNames as $newAttribute){
            $output->writeln("<info>Creating attribute: ".$newAttribute."</info>");
            $attribute = $this->objectManager->create('\Magento\Eav\Api\Data\AttributeInterface');
            $attribute->setEntityTypeId(4);
            $attribute->setAttributeCode($newAttribute);
            $attribute->setDefaultFrontendLabel(strtolower($newAttribute));
            $attribute->setFrontendInput('text');
            $attribute->setBackendType('varchar');
            $attribute->setIsRequired(false);

            try{
                $attribute = $this->attributeRepositoryInterface->save($attribute);
                $output->writeln("Attribute have attributeId: ".$attribute->getAttributeId());
                $atr = $this->attributeRepositoryInterface->get(4,$attribute->getAttributeCode());
                $output->writeln("The attribute saved with attributeCode ".$atr->getAttributeCode());
                $this->attributeManagementInterface->assign(
                    //$attributeSet->getEntityTypeId(),
                    \Magento\Catalog\Model\Product::ENTITY,
                    $attributeSet->getAttributeSetId(),
                    $attributeGroup->getAttributeGroupId(),
                    $attribute->getAttributeCode(),
                    1
                );
                $output->writeln("<info>Create attribute: ".$attribute->getAttributeCode()." and assigned to attriubuteGroup ".$attributeGroup->getAttributeGroupName()."</info>");
            } catch(\Exception $e) {
                var_dump($e->getFile());
                var_dump($e->getLine());
                var_dump($e->getTraceAsString());
                var_dump($e->getMessage());
            }
        }
    }
}
```
## Screenshot / Error Log

This is the error that compare during the execution of command and that i catch.
The principal error is "**Attribute with attributeCode "BRAND_UID" does not exist.**", it seems to be throws from get method of attributeRepository object.
![error](https://cloud.githubusercontent.com/assets/1780500/15173836/a30859fc-175e-11e6-8d23-a948afe96ebd.PNG)

I added these two lines of code just to debug , without these line of code the errors also appear because the same method (attributeRepository->get($entityTypeCode, $attributeCode)) is invoked by the assign method of the class that implements attributeManagementInterface.
This code can be deleted, was using only for debug.

```
$atr = $this->attributeRepositoryInterface->get(4,$attribute->getAttributeCode());
$output->writeln("The attribute saved with attributeCode ".$atr->getAttributeCode());
```
## Considerations 

At this time i have installed Magento 2.0.5, in a Virtual Machine with vagrant and ansible.
I have try with the _develop_ branch of Magento's repository but the error is the same.

Can anyone help me?
Thank you.

**Comment by maksek** [comment]
Hi @ilCleme, thank you for reporting your experience. It will take some time to go throw steps and then we will back to you.

**Comment by mattiaclementi** [comment]
Hi @maksek, thank you for your help.

I add some other information about the module i'm developing:
1. The module.xml is like this:

```
<?xml version="1.0"?>
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:module:Qweb_ImportProducts/etc/module.xsd">
    <module name="Qweb_ImportProducts" setup_version="1.0.0"/>
</config>
```

the module haven't dependency from other module. 
2. The di.xml contain only the injection of the command in the Magento CommandList:

```
<?xml version="1.0"?>
<!--
/**
 * Copyright © 2015 Magento. All rights reserved.
 * See COPYING.txt for license details.
 */
-->
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:ObjectManager/etc/config.xsd">
    <type name="Magento\Framework\Console\CommandList">
        <arguments>
            <argument name="commands" xsi:type="array">
                <item name="greeting_command" xsi:type="object">Qweb\ImportProducts\Console\Command\GreetingCommand</item>
                <item name="check_active_status_command" xsi:type="object">Qweb\ImportProducts\Console\Command\CheckActiveModulesCommand</item>
                <item name="create_product_command" xsi:type="object">Qweb\ImportProducts\Console\Command\CreateProductCommand</item>
            </argument>
        </arguments>
    </type>
</config>
```

There are many command inside this module, many of this command are only for example and exercise.
3. The composer.json contain this info:

```
{
  "name":"Qweb/ImportProducts",
  "type":"magento2-module",
  "autoload": {
    "psr-4": { "Qweb\\ImportProducts\\": "" },
    "files": [
      "registration.php"
    ]
  }
}
```

I don't set any dependency from other module.
I hope they can be useful information.

Thank you.

**Comment by magento-engcom-team** [comment]
@ilCleme, thank you for your report.
The issue is already fixed in develop branch, 2.1.10, 2.2.1
